# AUTHOR: Jon Tanner Nelson
# LAST UPDATED: 9-28-2022
# CONTENTS: write_csv_listOf_DFs(), consequential_mTable(),
# PURPOSE: Data retrieval and fixing functions which support the manipulation of GWAS catalog and Ensembl data.


#' write_csv_listOf_DFs
#'
#' iterates over a list of data.frames or table like data structures writing each to their own csv files.
#'
#' Names of files come from the names of the tables within the list input. Destination folder can be selected otherwise files will be exported into current wd.
#'
#' @param listOfDataFrames Data in. A list of data.frame(s) or similarly shaped objects which can be exported as CSVs.
#' @param destinationFolder folder the user wishes to export files to. By default files will export to current working directory
#' @param pattern specifies character or characters within data table names to replace by 'replacementChar'
#' @param replacementChar replacement character which will be inserted in place of 'patten' within file names
#' @param sortTables when TRUE will sort the tables according to the column specified by 'sortingCriteria'
#' @param sortingCriteria name of column which will be used to sort the data table by.
#' @param ascending will change the default sorting order from descending to ascending
#'
#' @returns
#' NA
#'
#' @example N/A
#'
#' @importFrom dplyr arrange
#' @importFrom dplyr desc
#' @importFrom stringr str_replace_all
#' @importFrom data.table fwrite
#'
#' @export
write_csv_listOf_DFs <- function(listOfDataFrames,
                                 destinationFolder = NA,
                                 pattern = NA,
                                 replacementChar = NA,
                                 sortTables = FALSE,
                                 sortingCriteria = 'variableName',
                                 ascending = FALSE){

  fileNames <- paste0(names(listOfDataFrames),'.csv')

  if(sortTables){ # Check for sortingCriteria Col before attempting sorting
    if( sum(names(listOfDataFrames[[1]]) %in% sortingCriteria) == 0){
      warning('The specified variable/(column) name cannot be found within the first Data Table, \nPlease enter a valid column name')
      stop()
    }
    if(ascending){ #sorting..
      sortedDataFrames <- lapply(listOfDataFrames, \(x) arrange(x, x[which( names(x) %in% sortingCriteria)] ) )
    } else {
      sortedDataFrames <- lapply(listOfDataFrames, \(x) arrange(x, desc( x[which( names(x) %in% sortingCriteria)] ) ) )
    }
    listOfDataFrames <- sortedDataFrames
  }

  if(!is.na(pattern)){ # edit fileNames.
    if(is.na(replacementChar)){
      warning('for replacement of \'pattern\' a replacementChar must be set')
      stop()
    }
    fileNames <- str_replace_all(fileNames, pattern, replacementChar)
  }

  if(!is.na(destinationFolder)){
    wd <- getwd()
    setwd(destinationFolder)
  }

  tryCatch(
    expr = {
      for(i in 1:length(listOfDataFrames)){
        fwrite(listOfDataFrames[[i]], file = fileNames[i])
      }
    },
    error = function(e){ # warning about file name issues, as they're common when writing files and typical warning messages aren't very informative.
      message('error encountered, check your file (list) names for invalid characters.')
      print(e)
    },
    warning = function(w){
      print(w)
    }
  )

  #resetting wd to its starting location before exiting function
  if(!is.na(destinationFolder)){
    setwd(wd)
  }
}


#' consequential_mTable
#'
#' selects columns which are useful for graphing. Also filters out variants which which are inconsequential according to their EnsVar_most_severe_consequence
#'
#' unfiltered master table goes in, narrowed down table comes out
#'
#' @param  mTable master Table generated by createMT() function
#'
#' @return
#' filtered data.table / tibble
#'
#' @example
#' graphableMT <- consequential_mTable(testMT[[1]])
#'
#' @importFrom dplyr select
#'
#' @noRd
consequential_mTable <- function(mTable) {

  mTableG <- select(mTable, VariantID, 'Reported trait', 'Trait(s)', 'P-value',
                    OR, CI, 'Mapped gene', Location, 'Study accession', EnsVar_MAF,
                    EnsVar_minor_allele, EnsVar_most_severe_consequence, EnsVar_allele_string)

  # pick out variant effects which are impactful and thus likely to be causative of some biological change
  impVarConseq <- c("regulatory_region_variant", "missense_variant","TF_binding_site_variant", "non_coding_transcript_exon_variant")

  hypotheticallyImportantRows <- mTableG[mTableG$EnsVar_most_severe_consequence %in% impVarConseq]
  return(hypotheticallyImportantRows)
}

# -------------------------------------------------------------------------






